tsdb/engine/tsm1/cache.go

```diff
@@ -0,0 +1,320 @@
+	return &entry{}
+}
+
+// add adds the given values to the entry.
+func (e *entry) add(values []Value) {
+	for _, v := range values {
+		// Only mark unsorted if not already marked.
+		if !e.unsorted && len(e.values) > 1 {
+			e.unsorted = e.values[len(e.values)-1].Time().UnixNano() >= v.Time().UnixNano()
+		}
+		e.values = append(e.values, v)
+		e.size += uint64(v.Size())
+	}
+}
+
+// dedupe remove duplicate entries for the same timestamp and sorts the entries.
+func (e *entry) dedupe() {
+	if !e.unsorted {
+		return
+	}
+	e.values = e.values.Deduplicate()
+	e.unsorted = false
+
+	// Update size.
+	e.size = 0
+	for _, v := range e.values {
+		e.size += uint64(v.Size())
+	}
+}
+
+type entries struct {
+	ee map[uint64]*entry
+}
+
+func newEntries() entries {
+	return entries{
+		ee: make(map[uint64]*entry),
+	}
+}
+
+func (a entries) add(values []Value, checkpoint uint64) {
+	e, ok := a.ee[checkpoint]
+	if !ok {
+		e = newEntry()
+		a.ee[checkpoint] = e
+	}
+	e.add(values)
+}
+
+// purge deletes all data that is as old as the checkpoint. Returns point-calculated
+// space freed-up.
+func (a entries) purge(checkpoint uint64) uint64 {
+	var size uint64
+	for k, v := range a.ee {
+		if k > checkpoint {
+			continue
+		}
+		size += v.size
+		delete(a.ee, k)
+	}
+	return size
+}
+
+// size returns point-calcuated storage size.
+func (a entries) size() uint64 {
+	var size uint64
+	for _, v := range a.ee {
+		size += v.size
+	}
+	return size
+}
+
+// clone returns the values for all entries under management, deduped and ordered by time.
+func (a entries) clone() Values {
+	var values Values
+	for _, v := range a.ee {
+		v.dedupe()
+		values = append(values, v.values...)
+	}
+	// XXX TO CONSIDER: it might be worth memoizing this.
+	return values.Deduplicate()
+}
+
+// Cache maintains an in-memory store of Values for a set of keys. As data is added to the cache
+// it will evict older data as necessary to make room for the new entries.
+type Cache struct {
+	mu         sync.RWMutex
+	store      map[string]entries
+	checkpoint uint64
+	size       uint64
+	maxSize    uint64
+
+	lru *lru // List of entry keys from most recently accessed to least.
+}
+
+// NewCache returns an instance of a cache which will use a maximum of maxSize bytes of memory.
+func NewCache(maxSize uint64) *Cache {
+	return &Cache{
+		maxSize: maxSize,
+		store:   make(map[string]entries),
+		lru:     newLRU(),
+	}
+}
+
+// WriteKey writes the set of values for the key to the cache. It associates the data with
+// the given checkpoint. This function is goroutine-safe.
+//
+// TODO: This function is a significant potential bottleneck. It is possible that keys could
+// be modified while an eviction process was taking place, so a big cache-level lock is in place.
+// Need to revisit this. It's correct but may not be performant (it is the same as the existing
+// design however)
+func (c *Cache) Write(key string, values []Value, checkpoint uint64) error {
+	c.mu.Lock()
```

Comment:
Oh I see why the other structs don't have mutexes now.  Big cache write-lock here.  Hmm.  I agree with your comments above.  This could create a write bottleneck, but probably fine for now until we have better data.